### PR TITLE
Automatically wrap large arguments with dask.delayed in map_blocks/partitions

### PR DESCRIPTION
**Pull request by mrocklin**
This avoids having many copies of an object in the graph, one in each task.

We should also do this in atop and in dd.map_partitions

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -2853,13 +2856,17 @@ def atop(func, out_ind, *args, **kwargs):
                              % (ind, arg.ndim))
 
     numblocks = {a.name: a.numblocks for a, ind in arginds if ind is not None}
-    argindsstr = list(concat([(a if ind is None else a.name, ind) for a, ind in arginds]))
+    argindsstr = list(concat([(delayed(a) if sizeof(a) > 1e6 else a
+                               if ind is None else a.name, ind)
```
Comment (jcrist) [inline dask/array/core.py]: I have to think about Python parsing rules to determine if this parses as

- `delayed(a) if sizeof(a) > 1e6 else (a if ind is None else a.name)`
-  `(delayed(a) if sizeof(a) > 1e6 else a) if ind is None else a.name`

I might put in the parens for how it's intended (even if unnecessary) just to make this clear on first glance.

Comment (mrocklin) [reply dask/array/core.py]: Fixed

**Comment by jakirkham** [comment]
So this is about the keyword arguments passed to `map_blocks` specifically, correct? If so, it might be worth noting this in the PR title (and the changelog eventually) instead of saying "objects" just for clarity. Just an observation 😄

**Comment by jakirkham** [comment]
Hmm...so I thought all the positional arguments were arrays, right? At least in the case of `map_blocks`. Is there a reason not to coerce the positional arguments to Dask Arrays in that case? Or am I missing something here?

**Comment by mrocklin** [comment]
They're not all necessarily arrays.  

```python
In [1]: import dask.array as da

In [2]: import operator

In [3]: x = da.ones(10, chunks=(5,))

In [4]: da.map_blocks(operator.add, x, 10).compute()
Out[4]: array([11., 11., 11., 11., 11., 11., 11., 11., 11., 11.])
```

**Comment by jakirkham** [comment]
Sure, though scalars could also be coerced to 0-D arrays.

**Comment by mrocklin** [comment]
I'm not currently seeing an advantage to that.  Note also that we might pass in arbitrary objects, like scikit-learn models or custom user defined types.

**Comment by jakirkham** [comment]
Not following. These other things are going in `map_blocks`'s positional arguments?

**Comment by mrocklin** [comment]
Yeah, you might imagine the following:

```python
def predict(data, model):
    return model.predict(data)

sgd = SGDClassifier(...)
sgd.fit(...)

da.map_blocks(predict, X, sgd)
```

**Comment by jakirkham** [comment]
Cool, well we might want to update the docstring then. Currently it only mentions arrays and scalars being supported.

**Comment by jakirkham** [comment]
Some other thoughts.

* Should we note the `delayed` values are pure?
* Should we check if the object is already lazy?

**Comment by mrocklin** [comment]
> Should we note the delayed values are pure?

We generally have a blanket assumption that the arguments you pass to Dask functions shouldn't change in place.  I don't think that we should repeat that in these methods' docstrings in particular.

> Should we check if the object is already lazy?

They're already small, so I don't think that this will come up, but it doesn't hurt to check.  Added.

**Comment by jakirkham** [comment]
>> Should we note the delayed values are pure?
>
> We generally have a blanket assumption that the arguments you pass to Dask functions shouldn't change in place.  I don't think that we should repeat that in these methods' docstrings in particular.

What I meant was using the `pure=True` argument with the `delayed` call.

**Comment by jakirkham** [comment]
>> Cool, well we might want to update the docstring then. Currently it only mentions arrays and scalars being supported.
>
> I just read through the docstring for map_blocks and atop and couldn't find this.  The parameters section says that this accepts arrays and constants.  That was the only mention that I found though.  Which section were you referring to?

Meant the parameters section in the docstring.

**Comment by mrocklin** [comment]
> Meant the parameters section in the docstring.

```python
def map_blocks(func, *args, **kwargs):
    """ Map a function across all blocks of a dask array.

    Parameters
    ----------
    func : callable
        Function to apply to every block in the array.
    args : dask arrays or constants
```

What is the language that you would like to change and what would you like to change it to?

**Comment by jakirkham** [comment]
[This example]( https://github.com/dask/dask/pull/4002#issuecomment-423671950 ) shows a use case that is not an array or a constant. How should we describe it? Are there other use cases that would be permitted and should be noted here somehow? Is this limited to `map_blocks` or does this affect other functions like `map_partitions`, which will also need this clarification?

**Comment by mrocklin** [comment]
In that comment I could call `sgd` a constant.  I suspect that we disagree on that term.  Can you recommend a better term?  Perhaps we should remove the "arrays or non-arrays" comment entirely.

**Comment by jakirkham** [comment]
Hmm...yeah, I would not call a mutable thing constant. 😉 When I read constant here, I think scalars (i.e. numeric values) in particular.

Stating dask arrays is helpful for noting they are handled differently. Perhaps just `objects`?

Ultimately my questions above are trying to get at what do *you* think should be allowable here. So I don't think having me answer the question is particularly useful for communicating the scope you intend as that is actually what I'm unclear about. 😄

**Comment by mrocklin** [comment]
You can place anything into map_blocks, arrays or otherwise, assuming that they aren't expected to change (as is true in all of dask).

I'm fine saying "arrays or other objects"

**Comment by mrocklin** [comment]
> Ultimately my questions above are trying to get at what do you think should be allowable here

I think that I've conveyed what is allowable above.  Mostly anything that isn't expected to mutate in the same way that any input to a dask function isn't expected to mutate.

> So I don't think having me answer the question is particularly useful

Well, currently I don't know how to satisfy what you're looking for, and I'd like to get this in.   I think that the current language is entirely fine.  Is it ok with you?  If not can I ask you to help me to understand how to improve it?

**Comment by mrocklin** [comment]
Is there anything I can do to move this along?  @jakirkham I would appreciate your help here.

**Comment by jakirkham** [comment]
Your proposed alternate wording sounds fine to me.

**Comment by mrocklin** [comment]
Done.  Pushed.  Merging.  Thanks @jakirkham !

**Pull request merged**